✨ Add support for extensions and updates to the OpenAPI schema in path operations

edouardlp opened the pull request:
## Description
This PR adds support for specifying extra info for the JSON schema of path operations. This is so that [OpenAPI extensions](https://swagger.io/docs/specification/openapi-extensions/) can be used when declaring routes.

## Use case

We need this because we rely on open api extensions to configure our service mesh (linkerd) for [retries](https://linkerd.io/2/tasks/configuring-retries/) and [timeouts](https://linkerd.io/2/tasks/configuring-timeouts/), but there are probably many use cases for this.

## Implementation
- There is a lot of boilerplate to add support for extra arguments, but it does not look like there is a way around it. Suggestions welcome.
- The pydantic model of `Operation` accepts any extra info, which matches the behavior of `Path`, `Field`, `Query`, etc. There might be an argument to restrict it to `x-` prefixed fields, but this will be more complex. I'm not sure it is worth it.

edouardlp committed:
fix(openapi): add support for extensions in path operations

edouardlp committed:
fix coverage

edouardlp commented:
Note that this is similar to https://github.com/tiangolo/fastapi/pull/1917/ in that `APIRoute` accepts extra arguments, but I believe the similarities end there. Both PRs appear compatible to me.

Kludex commented:
> Note that this is similar to #1917 in that APIRoute accepts extra arguments, but I believe the similarities end there. Both PRs appear compatible to me.

Right now, the PR's are not compatible... To make them compatible `extra` should be a `dict`, not a `keyworded` .
But as I said on that PR, I'm not sure if it will be accepted so I don't think you should addapt this PR (or be worried about compatibility) for the sake of that one. We should wait to see what @tiangolo has to say.

mbelang commented:
@tiangolo Any chance this could be looked at please? We think it is a nice enhancement to the openapi support and that would solve a big part of automation for us. Thx

triptec commented on docs_src/path_operation_advanced_configuration/tutorial005.py:
```diff
@@ -0,0 +1,8 @@
+app = FastAPI()
+
+
+@app.get("/items/", **{"x-my-open-api-extension": "value"})
```
@hadrien @edouardlp I think it should be openapi_extensions={..}, or if we'd use ** we should update the operation late in get_openapi_path instead of in get_openapi_operation_metadata as that would give us a "openapi override" so we could set requestBody or whatever and not having it overwritten somewhere in get_openapi_path.
What do you guys think?

triptec commented:
I tried this branch out and found a missing **route.extra in the call to add_api_route in include_router and also moved
```
            if route.extra:
                operation.update(route.extra)
```
to get_openapi_path.
I've tried to make PR to edouardlp:fix/support_openapi_extensions, but couldn't for some reason. The changes can be found at https://github.com/Future-Position-X/fastapi/commit/4d3f781bcf7bad93b41482bb24b5b29f39778b76

Is it perhaps something of interest?

tiangolo committed:
♻️ Update parameter name from wildcard to openapi_extra

to keep it comparable/consistent with Pydantic's schema_extra: https://pydantic-docs.helpmanual.io/usage/schema/#schema-customization

Remove wildcard parameters from path operation decorators, to be re-added as a single parameter after merging master

tiangolo committed:
🔀 Merge master

tiangolo committed:
♻️ Update implementation to extract openapi_extra, to allow extending the default

tiangolo committed:
✨ Add new parameter openapi_extra to FastAPI and APIRouter

tiangolo committed:
✅ Update test for extensions

tiangolo committed:
📝 Add additional source examples for extending the OpenAPI schema

tiangolo committed:
🍱 Add screenshot for OpenAPI extensions

tiangolo committed:
📝 Update docs for extending OpenAPI per path operation

tiangolo committed:
✅ Add tests for new openapi_extra examples

tiangolo commented:
Thanks for the work @edouardlp! :rocket: :cake: 

And thanks for the discussion everyone.

I updated this a bit, renamed the parameter to `openapi_extra`, to keep in line with [Pydantic's `schema_extra`](https://pydantic-docs.helpmanual.io/usage/schema/#schema-customization) while making it explicit that this is about OpenAPI.

I also refactored the implementation to make sure that it does a deep merge of the schema and moved it to do it after the other stuff is added. This way this can be used not only for extensions but also to modify the automatically generated schema.

I also added some other examples, docs, and tests.

This will be available in FastAPI version `0.68.0` in some minutes. 🔖🎉
